## Data Analysis Lab I

For this lab, you will continue work on your 4th TLAB. The dataset for this lab is in the `../data/shopping.csv` file. 

The goals for this process include:

* Generate new groups & bins to explore relationships
* Generate pivot tables for further analysis 
* Answer a few business-related questions

After you've completed this process, answer the questions underneath `Data Analysis Lab II` in your own words.

In [3]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
# TODO: load `../data/shopping.csv` as a pandas dataframe

df = pd.read_csv('../data/shopping.csv')

In [5]:
# TODO: Observe the head just to reorient ourselves with the dataframe. Take note of the columns and their types

df.head()

Customer ID   Age  Gender Item Purchased  Purchase Amount (USD)  Location  \
0         3475   NaN    Male         Jacket              30.904467     Maine   
1         3698  21.0  Female       Backpack              31.588259       NaN   
2         2756  31.0    Male       Leggings              24.231704    Nevada   
3         3340   NaN    Male        Pajamas              33.918834  Nebraska   
4         3391  38.0    Male     Sunglasses              36.545487    Oregon   

  Size         Color  Season  Review Rating Shipping Type Promo Code Used  \
0    M  Burnt orange    Fall            4.0      Standard              No   
1    L     Turquoise  Winter            2.0       Express              No   
2    M   Terra cotta  Winter            4.0      Standard              No   
3    M         Black  Winter            NaN      Standard              No   
4    S     Aubergine  Summer            NaN      Standard              No   

   Previous Purchases Payment Method Frequency of Purchases  
0                   0    Credit Card                    NaN  
1                   1    Credit Card                Monthly  
2                   0    Credit Card                    NaN  
3                   2    Credit Card              Bi-Weekly  
4                   0    Credit Card                    NaN

## Groups

In [6]:
# TODO: Group your dataframe according to the "Payment Method" column. Save this grouped dataframe into a new variable
# called 'g_pay' 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html

g_pay = df.groupby('Payment Method')

In [7]:
# TODO: Using the "agg" function calculate the mean and median of the "Purchase Amount (USD)" column in the `g_pay` object

g_pay['Purchase Amount (USD)'].agg(['mean', 'median'])

mean     median
Payment Method                      
Bank Transfer   42.724584  36.591483
Cash            37.174985  34.569802
Credit Card     38.607690  36.940467
Debit Card      38.826152  37.005280
PayPal          38.748823  36.708289
Venmo           38.595344  37.102529

In [8]:
# TODO: Using the "agg" function calculate the mean and median of the "Age" column in the `g_pay` object

g_pay['Age'].agg(['mean', 'median'])

mean  median
Payment Method                   
Bank Transfer   65.090909    64.0
Cash            65.300000    64.0
Credit Card     33.774270    33.0
Debit Card      33.595506    32.0
PayPal          35.355433    34.0
Venmo           34.505747    33.0

In [9]:
# TODO: Using the "agg" function calculate the mean and median of the "Review Rating" column in the `g_pay` object

g_pay['Review Rating'].agg(['mean', 'median'])

mean  median
Payment Method                  
Bank Transfer   4.000000     4.0
Cash            3.333333     4.0
Credit Card     3.874763     4.0
Debit Card      3.960334     4.0
PayPal          3.886792     4.0
Venmo           3.879808     4.0

In [10]:
# TODO: Group your dataframe according to the "Season" column. Save this grouped dataframe into a new variable
# called 'g_season' 

g_season = df.groupby(['Season'])

In [11]:
# TODO: Using the "agg" function calculate the mean and median of the "Purchase Amount (USD)" column in the `g_season` object

g_season['Purchase Amount (USD)'].agg(['mean', 'median'])

mean     median
Season                      
Fall    38.830085  36.770411
Spring  39.293314  37.866915
Summer  38.690885  36.843348
Winter  38.366148  36.691090

In [12]:
# TODO: Using the "agg" function calculate the mean and median of the "Age" column in the `g_season` object

g_season['Age'].agg(['mean', 'median'])

mean  median
Season                   
Fall    34.302083    32.0
Spring  33.772901    32.0
Summer  34.495887    34.0
Winter  34.262128    33.0

In [13]:
# TODO: Using the "agg" function calculate the mean and median of the "Review Rating" column in the `g_season` object

g_season['Review Rating'].agg(['mean', 'median'])

mean  median
Season                  
Fall    3.813953     4.0
Spring  3.892683     4.0
Summer  4.005882     4.0
Winter  3.911824     4.0

## Pivot Tables

In [14]:
# TODO: Create a pivot table where "Payment Method" is your column, "Shipping Type" is your index, and "Age" is your value
# use the "mean" function as the "aggfunc" parameter.
# Save this pivot table to a new variable named `tab_pay`
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html#

tab_pay = pd.pivot_table(df, columns = 'Payment Method', index = 'Shipping Type', values = 'Age', aggfunc = 'mean')
tab_pay

Payment Method  Bank Transfer       Cash  Credit Card  Debit Card     PayPal  \
Shipping Type                                                                  
2-Day Shipping      64.750000  63.000000    33.762274   33.925000  34.655629   
Express                   NaN        NaN    32.436170   32.957746  35.918919   
Next Day Air              NaN  69.000000    34.612500   33.431818  38.489362   
Standard            65.285714  65.428571    33.864023   33.525903  35.152597   
Store Pickup              NaN        NaN          NaN   19.000000        NaN   

Payment Method      Venmo  
Shipping Type              
2-Day Shipping  33.879195  
Express         32.833333  
Next Day Air    33.205128  
Standard        35.290210  
Store Pickup          NaN

In [15]:
# TODO: output this table for display. Notice the difference in average ages.

tab_pay

Payment Method  Bank Transfer       Cash  Credit Card  Debit Card     PayPal  \
Shipping Type                                                                  
2-Day Shipping      64.750000  63.000000    33.762274   33.925000  34.655629   
Express                   NaN        NaN    32.436170   32.957746  35.918919   
Next Day Air              NaN  69.000000    34.612500   33.431818  38.489362   
Standard            65.285714  65.428571    33.864023   33.525903  35.152597   
Store Pickup              NaN        NaN          NaN   19.000000        NaN   

Payment Method      Venmo  
Shipping Type              
2-Day Shipping  33.879195  
Express         32.833333  
Next Day Air    33.205128  
Standard        35.290210  
Store Pickup          NaN

In [16]:
# TODO: Create a pivot table where you analyze the count of purchases for each combination of "Color" and "Season"
# use the "count" function as the "aggfunc" parameter.
# Save this pivot table to a new variable named `tab_season`
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html#

tab_season = pd.pivot_table(df, columns = 'Season', index = 'Color', values = 'Item Purchased', aggfunc = 'count')

In [17]:
# TODO: Output this table for observation. Notice how counts of colors change throughout the season

tab_season

Season                Fall  Spring  Summer  Winter
Color                                             
Aubergine              101       4      14     144
Baby blue               17      83      99      24
Black                  103       5      12     155
Brick red              103       7      14     140
Brown                  117       5       5     120
Burnt orange           110       4      10     138
Fuchsia                  5      59     103      21
Lavender                19      63     120      13
Lemon yellow            11      62     112      18
Mauve                  102       6      12     120
Muted mustard yellow   103      10      15     110
Pale peach              13      68     104      15
Periwinkle               9      63     105       9
Ruby Red               112       4      10     132
Terra cotta            108       5      15     131
Turquoise               10      60      96      19
White                   13      73      96      12

In [18]:
# TODO: Using the "idmax" method, get the index with the maximum value for each Season column on the pivot table `tab_season`
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

tab_season.idxmax()

Season
Fall          Brown
Spring    Baby blue
Summer     Lavender
Winter        Black
dtype: object

## Bins 

In [19]:
# TODO: Create a new column called "purchase group" in your original dataframe which equally seperates your dataframe into 
# two groups of shoppers based on their "Purchase Amount (USD)" column using the "pd.qcut" method. 
# Ensure that you are only creating 2 "bins" and label these respective bins as ["Low", "High"]
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.qcut.html

df["purchase group"] = pd.qcut(x = df['Purchase Amount (USD)'], q = 2, labels = ['low', 'high'])

df["purchase group"]

0        low
1        low
2        low
3        low
4        low
        ... 
3895    high
3896    high
3897    high
3898    high
3899    high
Name: purchase group, Length: 3900, dtype: category
Categories (2, object): ['low' < 'high']

In [41]:
# TODO: Group your dataframe according to this new "purchase group" column. Save this grouped dataframe into a new variable
# called 'g_purchase' 

g_purchase = df.groupby(['purchase group'])
g_purchase



C:\Users\Uma\AppData\Local\Temp\ipykernel_34180\1817215103.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  g_purchase = df.groupby(['purchase group'])


In [57]:
# TODO: Using the "agg" function calculate the mean and median of the "Review Rating" column in the `g_purchase` object

g_purchase['Review Rating'].agg(['mean','median'])

 

mean  median
purchase group                  
low             3.862210     4.0
high            3.949857     4.0

In [59]:
# TODO: Using the "agg" function calculate the mean and median of the "Previous Purchases" column in the `g_purchase` object

g_purchase['Previous Purchases'].agg(['mean', 'median'])

mean  median
purchase group                  
low             5.984103     4.0
high            5.698974     4.0

## Hypothesis Testing 

Your manager has a hunch that people who have paid by credit card tend to spend more than people who have purchased by debit card. Let's test the statistical assumptions of this hypothesis.

What is the null hypothesis of this experiment? **Rememeber**: we are discussing the differences between average amounts spent between two groups.

H0: There are no differences in the spending habits between credit card users and debit card users. 

What is the alternative hypothesis of this experiment? **Rememeber**: we are discussing the differences between average amounts spent between two groups.

H1: Credit card users tend to spend more money than those making purchases with debit cards. 

In [68]:
# TODO: Create two new dataframes for only credit-card purchases and debit-card purchases, and save each respective dataframe into variables named `credit` and `debit` respectively. 
# Use boolean indexing to accomplish this. Recall that these values lie in the "Payment Method" column

credit = df[df['Payment Method'] == "Credit Card"]
credit


debit = df[df['Payment Method'] == "Debit Card"]
debit

Customer ID   Age  Gender Item Purchased  Purchase Amount (USD)  \
802          3105  18.0    Male         Hoodie              21.113629   
803          3441  51.0  Female          Scarf              30.722374   
804          3001  38.0  Female         Jacket              33.493934   
805          2379  19.0    Male       Leggings              38.263111   
806          2004  32.0  Female         Gloves              29.874618   
...           ...   ...     ...            ...                    ...   
3392          399  24.0    Male       Leggings              54.904077   
3393          734  29.0    Male        Pajamas              53.292856   
3394         1157   NaN    Male        T-shirt              45.638715   
3395          398  41.0    Male         Shorts              51.566554   
3396          294  32.0  Female          Scarf              54.446109   

           Location Size                 Color  Season  Review Rating  \
802         Georgia    S             Aubergine    Fall            5.0   
803        Maryland    L           Terra cotta    Fall            4.0   
804            Ohio    M                 Black  Winter            NaN   
805   New Hampshire    L  Muted mustard yellow    Fall            4.0   
806           Maine    S             Brick red    Fall            4.0   
...             ...  ...                   ...     ...            ...   
3392        Montana    S              Ruby Red  Winter            NaN   
3393            NaN    L  Muted mustard yellow    Fall            NaN   
3394         Hawaii    M              Lavender  Spring            4.0   
3395    Mississippi    M            Periwinkle  Spring            5.0   
3396    Connecticut    M              Ruby Red    Fall            NaN   

       Shipping Type Promo Code Used  Previous Purchases Payment Method  \
802         Standard              No                  13     Debit Card   
803         Standard              No                   1     Debit Card   
804         Standard              No                   1     Debit Card   
805     Next Day Air              No                  17     Debit Card   
806         Standard              No                   1     Debit Card   
...              ...             ...                 ...            ...   
3392  2-Day Shipping             Yes                  12     Debit Card   
3393  2-Day Shipping             Yes                   8     Debit Card   
3394  2-Day Shipping             Yes                   3     Debit Card   
3395        Standard             Yes                  15     Debit Card   
3396        Standard             Yes                   0     Debit Card   

     Frequency of Purchases purchase group  
802                Annually            low  
803               Quarterly            low  
804                 Monthly            low  
805                  Weekly           high  
806                 Monthly            low  
...                     ...            ...  
3392                    NaN           high  
3393                    NaN           high  
3394                    NaN           high  
3395                    NaN           high  
3396                    NaN           high  

[1300 rows x 16 columns]

In [69]:
# TODO: Calculate the p-value of the t-score between the "Purchase Amount (USD)" column of both these dataframes
# Documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

from scipy import stats 

stats.ttest_ind(credit["Purchase Amount (USD)"], debit["Purchase Amount (USD)"])

TtestResult(statistic=-0.4943296707098346, pvalue=0.6211136191569001, df=2698.0)

## Data Analysis Lab II

In the next section, answer a few questions regarding your dataset using the output you've generated.

## Q1

Based on the p-value that we've calculated, what can we state about the null hypothesis referring to the difference of average dollars spent between credit card and debit card holders.

Based on the p-value of 0.62, we fail to reject the the null hypothesis, stating that there is no difference between the credit and debit card group's spending habits. 

## Q2

What is the average age of a user who pays with cash? What is the average age of someone who pays with credit card? Is there a difference between these two users?

Sixety five is the average age of individuals who pay with cash, while 34 (rounded up) is the average age of individuals who pay with credit cards. There is a difference between groups, as those who pay with cash are much older than those who pay with credit card. 

## Q3

What is the average age of a user who buys clothes in the spring? What is the average age of a user who buys clothes in the winter Is there a difference between these two users?

The average age of people who buy clothes in the spring is 33 years old, while the average age of people who buy clothes in the winter is 34. There is not a big difference between the two groups. 

## Q4

What is the most popular season for the color "Lemon yellow"?

The Lemon yellow color sold mostly during the summer season. 

## Q5

What is the average review rating of a "low" purchase group user and the "high" purchase group?

The average review rating for the "Low" purchase group is 3.86, while the average review rating for the "High" purchase group is 3.94.